Ingest final data

vgeorge opened the pull request:
Opening this PR for visibility on the work of ingesting the final version of the data. This is a work in progress.

Current changes on the ingest workflow:

- Added instruction on how to copy the files from the Drive folder locally
- Fixed new folder structure
- Fixed new food group file ingest
- Fixed `runOgr2Ogr` function to allow spaces in filenames

The next step is to try to sucessfully ingest the smaller food group, which is Cloves. This process is failing now because the data files are separated by flow types, which is different from the previous approach. The mode is not included on the file as before. Header comparison:

- Previous: `from_id_admin,to_id_admin,flow_value,mode,segment_order,paths`
- Current: `from_id_admin,to_id_admin,flow_value,path_num,paths`

We need to clarify which mode should we assume for each file on the folder list:

- Flows/Land/Land_Domestic 
- Flows/Land/Land_ReExports
- Flows/Sea/Sea_Domestic 
- Flows/Land/Sea_ReExports
- Within_Country

cc @cameronkruse @kushankb @oliverroick @wrynearson

kushankb commented:
We couldn't include a modality in the final data because most of the flows consist of multiple modalities. For instance, a Sea/Sea_Domestic flow will consist of inland mode (rail, road or inland waterways) followed by maritime followed by inland mode again.
Flows/Land/Land_Domestic : can be a combination of rail, road, or inland waterways
Flows/Land/Land_ReExports:  can be a combination of rail, road, or inland waterways
Flows/Sea/Sea_Domestic: will be inland mode (rail, road or inland waterways)- maritime - inland mode
Flows/Land/Sea_ReExports: will be inland mode (rail, road or inland waterways)- maritime - inland mode
Within_Country: can be a combination of rail, road, or inland waterways

cameronkruse commented:
@vgeorge thanks for continuing to push on this. When you say, "We need to clarify which mode should we assume for each file" what are the implications of this for the app? Like Ginni said we don't have flow values any more, I'd recommend naming whatever flow value makes the rest of the app easiest to develop.

vgeorge commented:
@kushankb @cameronkruse thanks for the clarification. After looking more at the `Cloves` data, it seems the issues are exclusive to the `Land_ReExports` file:

- This file is the only one using a different header (the one I posted above); the others use the correct format
- *`mode` column*: To fix the lack of values, I made the script use `unknown`, which should be fine since this information is not used anywhere in the app at the moment
- *`path_num` column*: I added a change to convert this column into `segment_order`, which is the standard used in the other files

My current issue is that the `Land_ReExports/Flows_Cloves.csv` file includes a flow `IDN.11_1,NLD.8_1` that is already present in the `Land_Domestic/Flows_Cloves.csv` file. In our current schema, a flow must be unique. Can you please clarify why they are duplicated and how we should address it?

cameronkruse commented:
My guess on the duplicate naming is that the segment is used when cloves are exported from the country, but also, when cloves are exported after having been imported. I think, if that's the case we just group them together. @kushankb should give the final word on this though.

kushankb commented:
Hey @vgeorge,
The data is structured such that multiple types of flows can occur between any two administrative units. These flows include:
	•	Domestic flows by land
	•	Re-exports by land
	•	Domestic exports by sea
	•	Re-exports by sea

For example, food can flow between IDN.11_1 and NLD.8_1 via both land and sea, with flows categorized as either domestic exports or re-exports. Each of these flow types will have distinct segment order and paths.

It’s important to highlight that while multiple types of flows can exist between the same two admin units, there will not be duplicates within a single flow type. For instance, you won’t find duplicate “Land/Re-Exports” flows between IDN.11_1 and NLD.8_1.

Let me know if this explanation is clear! If not, I’d be happy to share additional illustrations or examples. Thanks!

vgeorge commented:
Hi Kushank, thanks for clarifying.

Why do Domestic and Reexport for the same food group use different routes between the source and destination?

In the Cloves example, they have different geometries depending on whether it is Domestic or Reexport.

At the moment, the [data schema](https://github.com/earth-genome/foodtwin-global-app/blob/develop/prisma/schema.prisma#L67-L77) does not allow more than one flow (A-B) between a pair of areas, other than by food group.

If I understood correctly, to accommodate this change, we would have to add a new property to the flow model to include the flow type Domestic or Reexport as a subgroup under a food group. Is that correct?

kushankb commented:
Hi @vgeorge, Happy New Year!

You're correct—the Domestic and Reexport flows will have different geometries. This is due to the nature of reexports: they involve food being exported from a region that doesn't produce them. For example, palm oil may be exported from the Netherlands to Germany, but since the Netherlands doesn’t produce palm oil (or enough palm oil), we need to trace the flow back to its primary source, such as Indonesia. As a result, reexport routes typically have more complex and longer geometries than domestic ones.

Therefore, there will be distinct geometries for different types of flows, including land-based domestic, land-based reexports, sea-based domestic, and sea-based reexports.

Let me know if that clarifies things, or if you need any further details.

vgeorge commented:
Hi @kushankb, happy new year for you too!

I’ll update the data schema and the ingest process to accommodate this change. I'll let you know if I have any questions on ingesting the data.

There are changes needed in other parts of the app to support the reexport flows, which require a broader discussion with the team to align on the implementation.

vgeorge commented:
Hi @cameronkruse, thank you for the video recording. I raised the issue because I looked into the pairs for IDN.11_1-NLD.8_1 in both `Land_Domestic` and `Land_ReExports` files, and the segments were not the same. Either way, even if the segments were the same, they are present in different flow type files, and an update to the schema is necessary. 

@kushank I updated the schema and the ingest script, but the issue I’m facing now is that some food groups have duplicated flows within the same file. Here’s a list of the food groups I tested:

- **Apples and products**:
    - Failed on AFG.17_1-NPL.1_1-73-LAND_REEXPORT
    - Failed on AFG.17_1-SAU.11_1-73-SEA_REEXPORT
    - Other files were ok
- **Aquatic Animals, Others**:
    - Failed on GRC.6_1-CHN.23_1-61-LAND_REEXPORT
    - Failed on CAN.11_1-CHN.10_1-61-SEA_REEXPORT
    - Other files were ok
- **Cloves**:
    - Failed on IDN.10_1-CAN.9_1-SEA_REEXPORT
    - Other files were ok

This doesn’t mean there aren’t other duplicated flows within the same file, the script skips the file after encountering the first duplicated flow.

I’m holding off on ingesting the data until we clarify this issue.

cc @wrynearson

kushankb commented:
Hi @vgeorge,
I looked into this, and it turns out there can be multiple flows between any two admin units within a food group, each with different geometries. This happens exclusively in the ReExport files, as we discussed this morning. These flows represent trade that has been traced back to the primary producing regions.

For example, in the case of Spain and Sweden, flows could either be direct (Spain → Sweden, domestic export flows) or pass through intermediate countries/admin units as ReExports. For instance, Spain → Germany → Sweden or Spain → Netherlands → Sweden. Each of these flows will have distinct geometries. In the clove example you flagged, there are two different flows between IDN.10_1-CAN.9_1 with separate geometries and flow values. One of these goes through EGT (IDN-->EGT-->CAN), while the other through IND (IDN-->IND-->CAN). 

While investigating, I noticed a discrepancy in the Sea_ReExports data that could inflate the flow values. I’m currently rerunning those flows to address this issue and will update the data folder on Google Drive over the weekend.

kushankb commented:
@vgeorge, please use the updated files for Sea_ReExports. They are in the same folder as before. Thanks!

vgeorge commented:
In order to accommodate the duplicate flows in the reexport described by @kushankb, the flow ID now [includes the flow value](https://github.com/earth-genome/foodtwin-global-app/blob/9833c3b937d612b42d3b42300eab8dd2cc8b53bb/prisma/seed/flows.ts#L223) to deduplicate flows between area pairs. This is a quick workaround to unblock the ingest process. This increases the required disk space but will allow us to ingest reexport data without big changes in the schema or the ingest script. After this, I didn't notice errors related to primary keys when ingesting Apples and Cloves.

I then tried to ingest some large food groups and they fail often due to lack of memory, so ingesting using a laptop doesn't look feasible. My next step is to spin up an instance on EC2 with (at least) 64GB of RAM to run the process on a more powerful machine.

cc @wrynearson @oliverroick

cameronkruse commented:
This sounds like a good work around @vgeorge do you think the larger file size will result in challenges rendering the data down the line? 

With regards to an EC2 instance. Do you have access to The Plotline Amazon instance?

vgeorge commented:
@cameronkruse regarding the ID change specifically, the impact is mostly on the ingest process and the required disk space, but it doesn't look significant. We’ll work on optimizations once the data is fully ingested.

As for the EC2 instance, I don’t believe I have access to the Plotline AWS instance, and it would be helpful to have it for the future. For now I’ll use Devseed’s instance to work on the ingest and let you know if it becomes a blocker.

vgeorge commented:
Today, I did some testing on using large EC2 instances, and the results were not very good. I first tried to use an `r7i.4xlarge`, but it was 4x slower than my local machine, which made me think the I/O was slower that on my laptop. I then created an `i4i.4xlarge` instance, which is optimized for I/O, but the results were only 25% faster when ingesting Cloves. I'm running Apples now, but it is not looking good, and it seems that throwing hardware at the problem won't help us. We need an ingest process that can run in a reasonable amount of time.

The main reason the files are so large is that edges have long IDs, like in this excerpt from a Clove:

```
ARE.2_1,ARE.7_1,5852,road,1,"['ARE.2_1-road14511185_ARE', 'road14511184_ARE-road14511185_ARE', 'road14511188_ARE-road14511184_ARE', 'road14511187_ARE-road14511188_ARE',...
```

In the edges table, we have about 7.4M edges, and if we used sequential integers for edge ids we could drastically reduce the flow file sizes and improve data movement between disk and memory during ingestion.

@kushankb do you think it is feasible to regenerate the flow files using this approach?

kushankb commented:
Hi @vgeorge, I don’t think that’s feasible for me at this point—it would require a significant effort, including fundamental changes to the pipeline and modifications to the IDs in the base infrastructure files.
Can you pre-process the long edges before ingestion? If so, you might be able to reduce the number of IDs within each path by aggregating consecutive road IDs. Since we don’t plan to visualize the exact road segments and are using some form of arcs, this approach could work well in reducing the size of the files. For instance, the clove paths above could be transformed into something like: 
['ARE.2_1-road14511185_ARE', .{road_arc}., port_id'].

vgeorge commented:
@kushankb thanks for the clarification, I understand it would be a big lift. I was able to find an alternate way.

I made the following updates to the code:

- Added sequential integer ID to the edges table
- Refactored edge and flow segment edge ID mapping to use integer IDs
- Replaced some CSV ingestion from temporary files with transaction inserts from memory
- Refactored the flow segment ID hash generator for shorter strings
- Disabled triggers during ingestion

These changes significantly reduced the memory footprint, and no memory errors occurred as a result. With the updates, I was able to ingest Apples in about 11 minutes, which is a more adequate ingest time. There are still optimization opportunities, but I don't think we need them right now. I'm running a full ingest and will report any errors once it’s finished.

vgeorge commented:
After some updated on the script I was able to do a full ingest locally. The database ended up with 385GB, so we will need to perform optimizations to have on the database only the data we need to serve the website. This can be merged now.
